### PR TITLE
feat: unsaved changes warning modal before route change

### LINKED ISSUE #10180: Warn for unsaved changes on event setup page.
### Issue Summary
Users doesn't get warning while leaving event setup page with unsaved changes.

### Actual Results
https://github.com/calcom/cal.com/assets/85894871/89d7ee73-5a3a-400e-9673-99cff07e8d4e

### Expected Results
Show a warning modal if user leaves without saving changes.

### PR DESCRIPTION
**Pull request by Krishna2323**
## What does this PR do?
Fixes #10180 

## Video Demo
https://github.com/calcom/cal.com/assets/85894871/1e6e2727-b4f3-4071-958a-3779211fae82

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?
1. Go to event type settings page
2. Make any changes
3. Try to change route by clicking any link on the sidebar

## Pseudocode 
1. Created a function to compare the default values and modified values.
2. Abort route change if there is any unsaved change in any editable tabs.
3. Save the aborted route to push after user saves or discard changes.

## Things left to fix
1. The warning modal title and message.

### CODE REVIEW

**Comment by Krishna2323** [comment]
Hey @PeerRich, can you pls check this.
I need help in fixing the modal, I can't add custom title or messages, that's something related to next-i18next, I think.

File: apps/web/pages/event-types/[type]/index.tsx [hunk 1]
```diff
@@ -132,6 +134,37 @@ export type FormValues = {
...
+  }
+
+  for (const key in obj2) {
+    if (obj2?.hasOwnProperty(key) && !obj1?.hasOwnProperty(key)) {
+      diff[key] = obj2[key];
+    }
+  }
+
+  return diff;
+}
```
Comment (sean-brydon) [inline apps/web/pages/event-types/[type]/index.tsx:166]: I havent tested but can we not use something like

const dirtyValues = watch({ dirtyFields: true });
from RHF to test for this?

File: apps/web/pages/event-types/[type]/index.tsx [hunk 2]
```diff
@@ -417,8 +468,42 @@ const EventTypePage = (props: EventTypeSetupProps) => {
...
+
+      if (shouldWarnUnsaved && !urlChanged) {
+        const unsavedChanges = getObjectDifference({ ...formMethods.getValues() }, defaultValues);
+
+        if (unsavedChanges["availability"]) {
+          delete unsavedChanges["availability"];
+        }
+
+        if (Object.keys(unsavedChanges).length > 0) {
+          console.log(unsavedChanges, formMethods.getValues(), defaultValues);
```
Comment (sean-brydon) [inline apps/web/pages/event-types/[type]/index.tsx:492]:
```suggestion
```

File: apps/web/pages/event-types/[type]/index.tsx [hunk 3]
```diff
@@ -519,6 +605,39 @@ const EventTypePage = (props: EventTypeSetupProps) => {
...
         />
       ) : null}
+
+      <Dialog open={unsavedDialogOpen} onOpenChange={setUnsavedDialogOpen}>
+        <ConfirmationDialogContent
+          isLoading={updateMutation.isLoading}
+          variety="danger"
+          cancelBtnText="Discard Changes"
+          // confirmBtnText="Keep Editing"
+          title={t(`delete${"_managed"}_event_type`)}
```
Comment (sean-brydon) [inline apps/web/pages/event-types/[type]/index.tsx:615]: Need to add the translation string to L615: place the string in common.json then use the key here

File: apps/web/pages/event-types/[type]/index.tsx [hunk 4]
```diff
@@ -519,6 +605,39 @@ const EventTypePage = (props: EventTypeSetupProps) => {
...
+          }}>
+          <p className="mt-5">
+            <Trans i18nKey="SaveChanges" components={{ li: <li />, ul: <ul className="ml-4 list-disc" /> }}>
+              <ul>
+                <li>Members assigned to this event type will also have their event types deleted.</li>
+                <li>
+                  Anyone who they&apos;ve shared their link with will no longer be able to book using it.
+                </li>
+              </ul>
+            </Trans>
```
Comment (sean-brydon) [inline apps/web/pages/event-types/[type]/index.tsx:637]:
```suggestion
            <Trans i18nKey="SaveChanges" components={{ li: <li />, ul: <ul className="ml-4 list-disc" /> }}>
              <ul>
                <li>Members assigned to this event type will also have their event types deleted.</li>
                <li>
                  Anyone who they&apos;ve shared their link with will no longer be able to book using it.
                </li>
              </ul>
            </Trans>
```
I think this text is wrong

**Review by sean-brydon — CHANGES_REQUESTED** [review]
Nice work - few NITs / suggestions i have left

File: apps/web/pages/event-types/[type]/index.tsx (hunk 3 above)
Comment (Krishna2323) [reply apps/web/pages/event-types/[type]/index.tsx:615]: @sean-brydon, which common.json? there are many.

File: apps/web/pages/event-types/[type]/index.tsx (hunk 4 above)
Comment (Krishna2323) [reply apps/web/pages/event-types/[type]/index.tsx:637]: fixed.

File: apps/web/pages/event-types/[type]/index.tsx (hunk 2 above)
Comment (Krishna2323) [reply apps/web/pages/event-types/[type]/index.tsx:492]: resolved.

File: apps/web/pages/event-types/[type]/index.tsx (hunk 1 above)
Comment (Krishna2323) [reply apps/web/pages/event-types/[type]/index.tsx:166]: I can't understand what you are saying :(

File: yarn.lock
[the remark this answers is not in the document]
Comment (Krishna2323) [reply yarn.lock:1]: this command doesn't remove the file :(

[the remark this answers is not in the document]
Comment (Udit-takkar) [reply yarn.lock:1]: i have removed the changes in yarn.lock

File: apps/web/pages/event-types/[type]/index.tsx (hunk 3 above)
Comment (Udit-takkar) [reply apps/web/pages/event-types/[type]/index.tsx:614]: "delete_managed_event_type" already exists in common.json.

@Krishna2323 if you have to add any new translation add it in locales/en/common.json.  you can also check i18next docs

File: apps/web/pages/event-types/[type]/index.tsx [hunk 5]
```diff
@@ -417,8 +468,41 @@ const EventTypePage = (props: EventTypeSetupProps) => {
...
+
+        if (Object.keys(unsavedChanges).length > 0) {
+          setUnsavedDialogOpen(true);
+          setNewRoute(url);
+          router.events.emit("routeChangeError");
+          throw "navigation aborted";
+        }
+      }
+    };
+    router.events.on("routeChangeStart", handleBeforeChange);
```
Comment (Udit-takkar) [inline apps/web/pages/event-types/[type]/index.tsx:499]: i am not sure if we should use router.events because the new next router doesn't support route events

File: apps/web/pages/event-types/[type]/index.tsx (hunk 1 above)
Comment (Udit-takkar) [reply apps/web/pages/event-types/[type]/index.tsx:166]: @Krishna2323 check this https://react-hook-form.com/docs/useformstate. React hook form gives us info about what fields have been changed in the form.

File: apps/web/public/static/locales/en/common.json [hunk 6]
```diff
@@ -1561,7 +1560,6 @@
   "apps_settings": "Apps settings",
   "fill_this_field": "Please fill in this field",
   "options": "Options",
-  "no_options_available":"No options available",
```
Comment (Udit-takkar) [inline apps/web/public/static/locales/en/common.json:1564]: why did you delete this?
